From 1896f6ac263fc7634aaf293590f77ffba81e05c6 Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:12:23 -0400
Subject: [virt] kvm/ppc: Add CAP to indicate hcall fixes

Message-id: <1410765214-16377-8-git-send-email-dgibson@redhat.com>
Patchwork-id: 94708
O-Subject: [PATCH 07/78] KVM: PPC: Add CAP to indicate hcall fixes
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1123145

Prerequisite to backport 9642382e826066c2d30d1b23d1b45410cdd8e07d

Signed-off-by: David Gibson <dgibson@redhat.com>

commit f2e91042a807cbf9b0b0d9776bf37d1ef0bd7ebe
Author: Alexander Graf <agraf@suse.de>
Date:   Thu May 22 17:40:15 2014 +0200

    KVM: PPC: Add CAP to indicate hcall fixes

    We worked around some nasty KVM magic page hcall breakages:

      1) NX bit not honored, so ignore NX when we detect it
      2) LE guests swizzle hypercall instruction

    Without these fixes in place, there's no way it would make sense to expose kvm
    hypercalls to a guest. Chances are immensely high it would trip over and break.

    So add a new CAP that gives user space a hint that we have workarounds for the
    bugs above in place. It can use those as hint to disable PV hypercalls when
    the guest CPU is anything POWER7 or higher and the host does not have fixes
    in place.

    Signed-off-by: Alexander Graf <agraf@suse.de>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

Conflicts:
 include/uapi/linux/kvm.h
---

diff --git a/arch/powerpc/kvm/powerpc.c b/arch/powerpc/kvm/powerpc.c
index a876a97..0410ca4 100644
--- a/arch/powerpc/kvm/powerpc.c
+++ b/arch/powerpc/kvm/powerpc.c
@@ -383,6 +383,7 @@ int kvm_dev_ioctl_check_extension(long ext)
  case KVM_CAP_SPAPR_TCE:
  case KVM_CAP_PPC_ALLOC_HTAB:
  case KVM_CAP_PPC_RTAS:
+ case KVM_CAP_PPC_FIXUP_HCALL:
 #ifdef CONFIG_KVM_XICS
  case KVM_CAP_IRQ_XICS:
 #endif
diff --git a/include/uapi/linux/kvm.h b/include/uapi/linux/kvm.h
index 720494d..37575be 100644
--- a/include/uapi/linux/kvm.h
+++ b/include/uapi/linux/kvm.h
@@ -678,6 +678,7 @@ struct kvm_ppc_smmu_info {
 #define KVM_CAP_IOAPIC_POLARITY_IGNORED 97
 #define KVM_CAP_ENABLE_CAP_VM 98
 #define KVM_CAP_IOEVENTFD_NO_LENGTH 100
+#define KVM_CAP_PPC_FIXUP_HCALL 103
 
 #ifdef KVM_CAP_IRQ_ROUTING
 
-- 
1.7.1